db/compaction/compaction_job.cc

```diff
@@ -290,6 +296,128 @@ void CompactionJob::Prepare() {
+  auto* c = compact_->compaction;
+  uint64_t max_subcompactions_limit =
+      static_cast<uint64_t>(c->max_subcompactions());
+  uint64_t num_planned_subcompactions =
+      std::min({static_cast<uint64_t>(ranges_size_limit),
+                max_subcompactions_limit, max_output_files_limit});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // For round-robin compaction prioity, we may have a large compaction with
+    // multiple input files in the start level. In this case, we need to employ
+    // more subcompactions (may exceed the max_subcompaction limit). The extra
+    // subcompactions will be executed using reserved threads and taken into
+    // account bg_compaction_scheduled. The number of planned subcompaction is
+    // determined through 3 steps:
+
+    // Step 1: Initialized against max_output_file_limit, the number of input
+    // files in the start level, and also the range size limit
+    num_planned_subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files_limit,
+         static_cast<uint64_t>(ranges_size_limit)});
+
+    TEST_SYNC_POINT_CALLBACK("CompactionJob::GetNumPlannedSubcompactions:Init",
+                             &num_planned_subcompactions);
+    if (max_subcompactions_limit < num_planned_subcompactions) {
+      // Step 2: Check if the initial number exceeds user-configured
+      // max_subcompaction. If yes, we need to compute the extra subcompaction
+      // based on compaction limits (the extra subcompactions should not
+      // be larger than that)
+      int extra_num_subcompactions_limit =
+          std::min((int)(num_planned_subcompactions - max_subcompactions_limit),
+                   available_bg_compactions_db_limit_);
+
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:"
+          "ExtraCompactionsAgainstDBLimits",
+          &extra_num_subcompactions_limit);
+
+      // Step 3: Try to reserve threads for extra subcompactions. The planned
+      // subcompactions may change again if we cannot obtain sufficient threads
+      extra_num_subcompaction_threads_reserved_ =
+          env_->ReserveThreads(extra_num_subcompactions_limit, thread_pri_);
+      num_planned_subcompactions =
+          static_cast<uint64_t>(extra_num_subcompaction_threads_reserved_) +
+          max_subcompactions_limit;
+      TEST_SYNC_POINT_CALLBACK(
+          "CompactionJob::GetNumPlannedSubcompactions:Finalize",
+          &num_planned_subcompactions);
+    }
+  }
+  return num_planned_subcompactions;
+}
+
+void CompactionJob::UpdateResourcesAndLimits(
+    uint64_t num_actual_subcompactions, uint64_t num_planned_subcompactions) {
+  // We only need to update resources and limits if the actual subcompactions
+  // is less than the planned one and we did reserve some resources earlier
+  if (extra_num_subcompaction_threads_reserved_ > 0 &&
+      num_actual_subcompactions < num_planned_subcompactions) {
```

Comment:
I believe we can move this out of `CompactionJob::UpdateResourcesAndLimits` and call the function conditionally to reduce nested if-else within this function.

I also wonder if extra_num_subcompaction_threads_reserved_ > 0 is needed. Would `extra_num_subcompaction_threads_reserved_` ever be negative? If it is 0, std::min should takes care of it and do env_->ReleaseThreads(0).

Reply:
`extra_num_subcompaction_threads_reserved_` can never be negative. It seems that I forgot to add a condition in `ReserveThreads()` and `ReleaseThreads()` to add the mutex only if the input number is not zero. For now, even the input of `ReleaseThreads` is 0, we still have mutex and `WakeUpAllThreads()` once. Should I change them after removing `extra_num_subcompaction_threads_reserved_ > 0`?

Reply:
> It seems that I forgot to add a condition in ReserveThreads() and ReleaseThreads() to add the mutex only if the input number is not zero. For now, even the input of ReleaseThreads is 0, we still have mutex and WakeUpAllThreads() once.

Does that mean the `ReserveThreads()/ReleaseThreads()` still have some performance overhead when the threads = 0 without any correctness issue? If it is, it sounds to me that the performance overhead is acceptable and you may fix that as a follow-up PR before or after this PR lands.

Reply:
Yeah - then it's fine to keep the `extra_num_subcompaction_threads_reserved_ > 0` in the if(...)

Reply:
I think it should not be possible to provide a larger number to Shrink than what was provided to or returned by Reserve. To do that you can make the meaning of the Shrink argument similar to Reserve in the sense that it accounts for the subcompactions that might be running without resources (per `max_subcompactions`). Practically, that just means moving the `std::min(extra_num_subcompaction_threads_reserved_, num_unnecessary_subcompactions)` into the caller of Shrink, I think.

Reply:
In that case both the conditions surrounding the caller (`extra_num_subcompaction_threads_reserved_ > 0` and `num_actual_subcompactions < num_planned_subcompactions`) can be replaced by checking whether the Shrink argument is zero. Personally I'd make sure Shrink can support shrinking by zero and allow the call to be made.